### ISSUE TITLE
Integrate Error Correction Code Transformer (ECCT) for Neural Decoding

### ISSUE THREAD

**Issue by selimfirat** [ipc-lab/kaira#35]
# Integrate Error Correction Code Transformer (ECCT) for Neural Decoding

## Summary

Integrate the Error Correction Code Transformer (ECCT) architecture from ["Error Correction Code Transformer" (NeurIPS 2022)](https://arxiv.org/abs/2203.14966) into Kaira as a state-of-the-art neural decoder for linear error correction codes.

## Background

The Error Correction Code Transformer represents a breakthrough in neural decoding by extending Transformer architecture to soft decoding of linear codes at arbitrary block lengths. This approach outperforms existing neural decoders while addressing key limitations of previous methods.

### Key Innovation: Masked Self-Attention for ECC

ECCT introduces a novel **masked self-attention mechanism** specifically designed for error correction codes:
- **Variable Node Connections**: Bits connected through parity check equations can attend to each other
- **Check Node Integration**: Syndrome information is integrated through attention between check nodes and connected variable nodes
- **Structural Sparsity**: The attention mask reflects the code's parity check matrix structure, making computation efficient

### Advantages over Existing Neural Decoders

1. **No Belief Propagation Dependency**: Unlike neural BP variants, ECCT doesn't rely on iterative message passing
2. **Reduced Overfitting**: Avoids exponential training complexity of exhaustive neural approaches  
3. **Flexible Architecture**: Works with arbitrary block lengths and code types
4. **Superior Performance**: Outperforms state-of-the-art neural decoders by large margins
5. **Computational Efficiency**: Fraction of time complexity compared to existing methods

#### `kaira/models/fec/decoders/ecct_decoder.py`
```python
class ECCTransformerDecoder(nn.Module):
    """
    Error Correction Code Transformer Decoder
    
    Implements the ECCT architecture for neural decoding of linear codes
    using masked self-attention based on parity check matrix structure.
    """
    
    def __init__(self, 
                 parity_check_matrix: torch.Tensor,
                 d_model: int = 64,
                 num_layers: int = 6, 
                 num_heads: int = 8,
                 dropout: float = 0.1):
        """
        Args:
            parity_check_matrix: Binary parity check matrix (m x n)
            d_model: Model dimension for embeddings
            num_layers: Number of transformer encoder layers
            num_heads: Number of attention heads
            dropout: Dropout probability
        """
        
    def forward(self, 
                received_soft: torch.Tensor,
                syndrome: torch.Tensor) -> torch.Tensor:
        """
        Args:
            received_soft: Soft channel outputs (batch_size, n)
            syndrome: Computed syndrome (batch_size, m) 
            
        Returns:
            Decoded bits (batch_size, n)
        """
```

#### Transformer Components
- **ECCTransformerLayer**: Encoder layer with masked self-attention
- **ECCMultiHeadAttention**: Multi-head attention with ECC-specific masking
- **PositionwiseFeedForward**: Feed-forward network with GELU activation
- **ECCEmbedding**: Learnable embeddings for variable and check nodes

#### Attention Mask Generation
```python
def build_ecc_attention_mask(parity_check_matrix: torch.Tensor) -> torch.Tensor:
    """
    Build attention mask based on parity check matrix structure.
    
    - Variable nodes connected by same check equation can attend to each other
    - Check nodes can attend to their connected variable nodes
    - Self-attention always allowed (diagonal)
    
    Returns sparse attention mask reflecting code structure.
    """
```

#### Decoder Interface Compatibility
```python
# Inherit from existing BeliefPropagationDecoder interface
class ECCTransformerDecoder(BaseDecoder):
    def __init__(self, encoder: LDPCCodeEncoder, **kwargs):
        self.parity_check_matrix = encoder.check_matrix
        # Initialize transformer architecture
        
    def decode(self, received_soft: torch.Tensor) -> torch.Tensor:
        # Compute syndrome
        syndrome = self.compute_syndrome(received_soft)
        # Apply transformer decoding
        return self.forward(received_soft, syndrome)
```

#### Training Integration
```python
# Support for various code types already in Kaira
def train_ecct_decoder(code_type: str, 
                       code_params: Dict,
                       training_config: Dict):
    """
    Train ECCT decoder for specified code type.
    
    Supports:
    - LDPC codes (from RPTU database or custom)
    - Polar codes  
    - BCH codes
    - Custom linear codes
    """
```

#### LDPC Codes (Priority 1)
- Integrate with existing `LDPCCodeEncoder` 
- Support RPTU database codes (WiMAX, WiFi, 5G, etc.)
- Custom LDPC matrices

#### Polar Codes (Priority 2)  
- Support standard polar code constructions
- Integration with successive cancellation decoding comparison

#### BCH Codes (Priority 3)
- Classical BCH codes for comparison with algebraic decoding

#### Custom Linear Codes (Priority 4)
- Any linear code with known parity check matrix

#### Training Example
```python
# examples/models_fec/train_ecct_decoder.py
"""
Train Error Correction Code Transformer for LDPC decoding.

Demonstrates:
- ECCT decoder setup for WiMAX LDPC codes
- Training with AWGN channel simulation  
- Performance comparison with BP decoding
- BER/FER evaluation across SNR range
"""

import torch
from kaira.models.fec.encoders import LDPCCodeEncoder
from kaira.models.fec.decoders import ECCTransformerDecoder, BeliefPropagationDecoder
from kaira.channels.analog import AWGNChannel
from kaira.metrics.signal import BitErrorRate, FrameErrorRate

# Create LDPC encoder (WiMAX standard)
encoder = LDPCCodeEncoder(
    rptu_database=True, 
    code_length=576, 
    code_dimension=288, 
    rptu_standart="wimax"
)

# Initialize ECCT decoder
ecct_decoder = ECCTransformerDecoder(
    encoder=encoder,
    d_model=64,
    num_layers=6,
    num_heads=8
)

# Training loop with AWGN channel
for epoch in range(num_epochs):
    for batch in dataloader:
        messages = batch['messages']
        codewords = encoder(messages)
        
        # Add AWGN noise
        channel = AWGNChannel(snr_db=snr)
        received = channel(codewords)
        
        # ECCT decoding
        decoded = ecct_decoder(received)
        
        # Compute loss and backpropagate
        loss = ecct_decoder.loss(decoded, messages, received)
        loss.backward()
        optimizer.step()
```

#### Performance Comparison Example
```python
# examples/models_fec/compare_ecct_vs_bp.py
"""
Performance comparison: ECCT vs Belief Propagation decoding.

Shows superior performance of neural ECCT decoder while
maintaining educational value of understanding both approaches.
"""

def compare_decoders():
    # Test both ECCT and BP decoders
    ecct_ber = evaluate_decoder(ecct_decoder, test_data)
    bp_ber = evaluate_decoder(bp_decoder, test_data) 
    
    # Plot comparison showing ECCT advantages
    plot_ber_comparison(ecct_ber, bp_ber)
```

### Phase 1: Core Architecture (2-3 weeks)
- [ ] Implement `ECCTransformerDecoder` base class
- [ ] Create transformer layer components (attention, feed-forward)
- [ ] Implement attention mask generation from parity check matrices
- [ ] Add syndrome computation utilities
- [ ] Unit tests for architecture components

### Phase 2: Training Framework (2 weeks)  
- [ ] Integrate with existing Kaira training infrastructure
- [ ] Add ECCT-specific loss functions and optimizers
- [ ] Support for different code types (LDPC, Polar, BCH)
- [ ] Training utilities and data loaders
- [ ] Integration tests with existing encoders

### Phase 3: Examples and Documentation (1-2 weeks)
- [ ] Create comprehensive training examples
- [ ] Performance comparison examples (ECCT vs BP)
- [ ] Tutorial notebooks for different code types
- [ ] API documentation and usage guides
- [ ] Benchmark results and performance analysis

### Phase 4: Advanced Features (2 weeks)
- [ ] Support for punctured and shortened codes
- [ ] Mixed-precision training optimization
- [ ] Distributed training support
- [ ] Model compression techniques
- [ ] Advanced attention mechanisms (sparse patterns)

### Attention Mask Construction
The key innovation is building attention masks that reflect the code structure:
```python
def build_attention_mask(parity_check_matrix):
    n, m = parity_check_matrix.shape  # n=block length, m=check nodes
    total_size = n + m  # variable nodes + check nodes
    
    mask = torch.eye(total_size)  # Self-attention always allowed
    
    # For each parity check equation
    for check_idx in range(m):
        # Get variable nodes connected to this check
        connected_vars = torch.where(parity_check_matrix[check_idx] > 0)[0]
        
        # Allow attention between connected variables  
        for i in connected_vars:
            for j in connected_vars:
                if i != j:
                    mask[i, j] = 1  # var-var attention
                    
            # Allow attention between check and variables
            mask[n + check_idx, i] = 1  # check-var attention
            mask[i, n + check_idx] = 1  # var-check attention
                    
    return ~mask.bool()  # Invert for attention masking
```

### Model Architecture
```
Input: [magnitude(n), syndrome(m)] → Embedding(n+m, d_model)
  ↓
Transformer Encoder (L layers):
  - Masked Multi-Head Self-Attention  
  - Position-wise Feed-Forward
  - Layer Normalization & Residual Connections
  ↓  
Output Projection: (n+m, d_model) → (n,) 
  ↓
Sigmoid → Decoded Bits
```

### Loss Function
Binary cross-entropy with logits, optimized for bit-level accuracy:
```python
def ecct_loss(logits, target_bits, received_soft):
    # Convert to probabilities 
    bit_probs = torch.sigmoid(logits)
    
    # Binary cross-entropy loss
    bce_loss = F.binary_cross_entropy_with_logits(logits, target_bits)
    
    # Optional: Add syndrome consistency loss
    syndrome_loss = syndrome_consistency_penalty(bit_probs, parity_matrix)
    
    return bce_loss + syndrome_loss
```

### Research Benefits
1. **State-of-the-Art Performance**: Access to cutting-edge neural decoding
2. **Flexible Experimentation**: Easy to modify architecture for research
3. **Comparative Studies**: Direct comparison with BP and other methods
4. **Code Agnostic**: Works with any linear code type

### Educational Benefits  
1. **Modern Architecture Understanding**: Learn transformer applications beyond NLP
2. **Attention Mechanism Insights**: See how attention applies to graph-structured data
3. **Neural vs Classical**: Compare neural and classical decoding approaches
4. **Implementation Deep-Dive**: Understand transformer internals in new domain

### Practical Benefits
1. **Performance Gains**: Significant improvement over classical methods
2. **Deployment Ready**: Trained models can be exported for production
3. **Efficient Inference**: Optimized attention patterns reduce computation
4. **Scalable**: Works across different code lengths and rates

## Success Metrics

1. **Performance**: ECCT decoder achieves reported performance gains vs BP
2. **Compatibility**: Seamless integration with existing Kaira encoders  
3. **Usability**: Clear examples and documentation for easy adoption
4. **Efficiency**: Training and inference times competitive with original implementation
5. **Extensibility**: Easy to extend for new code types and research directions

## References

- **Paper**: [Error Correction Code Transformer](https://arxiv.org/abs/2203.14966) (NeurIPS 2022)
- **Original Code**: [ECCT GitHub Repository](https://github.com/yoniLc/ECCT)
- **Authors**: Yoni Choukroun, Lior Wolf

## Implementation Notes

- Maintain compatibility with existing Kaira decoder interface
- Leverage existing infrastructure for channel simulation and metrics
- Ensure proper documentation and examples for educational use
- Consider computational requirements and provide optimization guidance
- Support both research and production use cases

This integration will significantly enhance Kaira's capabilities by adding state-of-the-art neural decoding while maintaining the library's focus on education, research, and practical applications.
